File: _java-archive/warranty-service/src/main/java/dev/kritchalach/warranty/domain/CoveragePeriod.java
package dev.kritchalach.warranty.domain;

import jakarta.persistence.Embeddable;
import java.time.Instant;

/**
 * Value Object: CoveragePeriod
 * ช่วงเวลาคุ้มครอง warranty (start, end) immutable
 */
@Embeddable
public class CoveragePeriod {

    private Instant startsAt;
    private Instant endsAt;

    protected CoveragePeriod() {}

    public CoveragePeriod(Instant startsAt, Instant endsAt) {
        if (startsAt == null || endsAt == null) {
            throw new IllegalArgumentException("CoveragePeriod requires both startsAt and endsAt");
        }
        if (!endsAt.isAfter(startsAt)) {
            throw new IllegalArgumentException("endsAt must be after startsAt");
        }
        this.startsAt = startsAt;
        this.endsAt = endsAt;
    }

    public boolean covers(Instant moment) {
        return !moment.isBefore(startsAt) && !moment.isAfter(endsAt);
    }

    public Instant getStartsAt() { return startsAt; }
    public Instant getEndsAt() { return endsAt; }
}

File: warranty-service/src/domain/CoveragePeriod.ts
/**
 * Value Object: CoveragePeriod
 * ช่วงเวลาคุ้มครอง warranty (start, end) immutable
 */
export class CoveragePeriod {
  readonly startsAt: string;
  readonly endsAt: string;

  constructor(startsAt: string, endsAt: string) {
    if (!startsAt || !endsAt) {
      throw new Error('CoveragePeriod requires both startsAt and endsAt');
    }
    if (new Date(endsAt).getTime() <= new Date(startsAt).getTime()) {
      throw new Error('endsAt must be after startsAt');
    }
    this.startsAt = startsAt;
    this.endsAt = endsAt;
  }

  covers(momentIso: string): boolean {
    const m = new Date(momentIso).getTime();
    return m >= new Date(this.startsAt).getTime() && m <= new Date(this.endsAt).getTime();
  }
}
